util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -74,195 +93,217 @@ var (
-func (w *Worker) StartTCPServer(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting TCP Server")
-	w.startWork(rw, request, ProtocolTCP, "iperf", tcpServerArguments)
-}
-
-// StartUDPServer starts iperf server for udp measurements.
-func (w *Worker) StartUDPServer(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting UDP Server")
-	w.startWork(rw, request, ProtocolUDP, "iperf", udpServerArguments)
-}
-
-// StartHTTPClient starts an http client for http measurements.
-func (w *Worker) StartHTTPClient(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting HTTP Client")
-	w.startWork(rw, request, ProtocolHTTP, "siege", httpClientArguments)
+func (w *Worker) populateResourceSpec(object runtime.Object) error {
+	var err error
+	w.work.resourceContent, err = runtime.DefaultUnstructuredConverter.ToUnstructured(object)
+	if err != nil {
+		return fmt.Errorf("error converting event Object to unstructured.Event object: %s", object)
+	}
+	w.work.resourceSpecMap = w.work.resourceContent["spec"].(map[string]interface{})
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(w.work.resourceSpecMap, &w.work.resourceSpec); err != nil {
+		return fmt.Errorf("error converting custom resource properties: %s", err)
+	}
+	return nil
 }
 
-// StartTCPClient starts iperf client for tcp measurements.
-func (w *Worker) StartTCPClient(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting TCP Client")
-	w.startWork(rw, request, ProtocolTCP, "iperf", tcpClientArguments)
+func (w *Worker) startWork() {
+	if w.work.resourceSpec.Protocol == ProtocolHTTP && w.work.workType == "server" {
+		w.StartHTTPServer()
+		return
+	}
+	var err error
+	properties := extractParameters(w.work.resourceSpecMap)
+	w.work.arguments, err = populateTemplates(protocolArgumentMap[w.work.workType][w.work.resourceSpec.Protocol], properties)
+	if err != nil {
+		err = fmt.Errorf("populating template failed: %v", err)
+		klog.Error(err)
+		w.updateStatus(makeErrorStatus(err))
+		return
+	}
+	klog.Infof("Populated templates: %s", w.work.arguments)
+	startTimestamp := w.getStartTimestamp()
+	w.schedule(startTimestamp)
 }
 
-// StartUDPClient starts iperf client for udp measurements.
-func (w *Worker) StartUDPClient(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting UDP Client")
-	w.startWork(rw, request, ProtocolUDP, "iperf", udpClientArguments)
+func (w *Worker) getStartTimestamp() int64 {
+	if w.work.workType == "client" {
+		return w.work.resourceSpec.ClientStartTimestamp
+	}
+	return time.Now().Unix()
 }
 
-// Handler handles http requests for http measurements.
-func (w *Worker) Handler(rw http.ResponseWriter, request *http.Request) {
-	w.sendResponse(rw, http.StatusOK, "ok")
+func (w *Worker) updateStatus(status map[string]interface{}) {
+	w.work.resourceContent["status"] = status
+	var unstructuredRes unstructured.Unstructured
+	unstructuredRes.SetUnstructuredContent(w.work.resourceContent)
```

Comment:
Instead of modifying resourceContent etc., let's simply use the 
UpdateStatus method, instead of Update:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/dynamic/interface.go#L36

Reply:
As informed ,have issues with UpdateStatus,hence for now not changing this.

Reply:
I guess I wasn't clear offline. What I said is that, instead of storing "resourceContent" explicit in the Worker just do
(syntax to fix):

obj, err := w.resourceInterface.Get(context.TODO(), name)
// handle err
// transform to map[string]interface{}
// do what you do here

This let's you remove resourceContent from Worker struct too (without doing UpdateStatus).